core/src/main/java/org/mule/runtime/core/internal/connection/DefaultConnectivityTesterFactory.java

```diff
@@ -0,0 +1,178 @@
+
+import static java.lang.Boolean.valueOf;
+import static java.lang.Integer.getInteger;
+import static java.lang.String.format;
+import static java.lang.System.getProperty;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.api.util.MuleSystemProperties.ASYNC_TEST_CONNECTIVITY_TIMEOUT_PROPERTY;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.connection.ConnectionProvider;
+import org.mule.runtime.api.connection.ConnectionValidationResult;
+import org.mule.runtime.api.exception.DefaultMuleException;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lock.LockFactory;
+import org.mule.runtime.api.scheduler.Scheduler;
+import org.mule.runtime.api.scheduler.SchedulerService;
+import org.mule.runtime.api.util.LazyValue;
+import org.mule.runtime.api.util.concurrent.Latch;
+import org.mule.runtime.core.api.retry.RetryCallback;
+import org.mule.runtime.core.api.retry.RetryContext;
+import org.mule.runtime.core.api.retry.policy.RetryPolicyTemplate;
+import org.mule.runtime.core.api.util.func.CheckedRunnable;
+import org.mule.runtime.core.internal.retry.ReconnectionConfig;
+import org.mule.runtime.extension.api.connectivity.NoConnectivityTest;
+import org.mule.runtime.extension.api.runtime.config.ConfigurationInstance;
+
+import java.util.concurrent.locks.Lock;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+
+/**
+ * Implementation of {@link ConnectivityTesterFactory} that takes into account the reconnection strategy.
+ *
+ * @since 4.4
+ */
+public class DefaultConnectivityTesterFactory implements ConnectivityTesterFactory {
+
+  public static final String DO_TEST_CONNECTIVITY_PROPERTY_NAME = "doTestConnectivity";
+
+  private static final Integer DEFAULT_ASYNC_TEST_CONNECTIVITY_TIMEOUT = 30000;
+
+  private static final int ASYNC_TEST_CONNECTIVITY_TIMEOUT =
+      getInteger(ASYNC_TEST_CONNECTIVITY_TIMEOUT_PROPERTY, DEFAULT_ASYNC_TEST_CONNECTIVITY_TIMEOUT);
+
+  private static final Logger LOGGER = getLogger(DefaultConnectivityTesterFactory.class);
+
+  @Inject
+  private ConnectionManagerAdapter connectionManager;
+
+  @Inject
+  private LockFactory lockFactory;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private final LazyValue<Boolean> doTestConnectivity = new LazyValue<>(this::getDoTestConnectivityProperty);
+
+  @Override
+  public ConnectivityTester create(String name) {
+    final Lock testConnectivityLock = lockFactory.createLock(this.getClass().getName() + "-testConnectivity-" + name);
+    return new ConnectivityTester() {
+
+      @Override
+      public void testConnectivity(ConnectionProvider provider, ConfigurationInstance configurationInstance)
+          throws MuleException {
+        if (provider instanceof NoConnectivityTest || !doTestConnectivity.get()) {
```

Comment:
Is there a way to avoid instanceof without changing api? It becomes a problem if some day we want to decorate the provider.

Reply:
The ideal would be to avoid the if entirely with some kind of double dispatch but maybe we can avoid the instanceof and avoid problems if some day we have to wrap the ConnectionProvider.

Reply:
`NoConnectivityTest` comes from the sdk, not runtime, so cannot do that without extending its API which would have a much broader impact.

Reply:
Actually, seeing how this interface is used in the connectors, an annotation would have been better. But then again, this logic in the runtime would be very similiar. Keep in mind that the design entention of the sdk is to make life easier to the people writing connectors, not the runtime.

Reply:
Ok. Even if this is used by the connectors the best way would not to depend of the interface with an instanceof to decide one thing or the other. This cannot be changed for this PR but should be taken into account.